risc0/zkvm/src/prove/io.rs

```diff
@@ -123,16 +122,142 @@ impl<H: SliceIoHandler> RawIoHandler for SliceIoWrapper<H> {
     }
 }
 
+/// Generates a Syscall from a simple slice function
+pub fn slice_io_from_fn<T: Pod, U: Pod, F: Fn(&[T]) -> Vec<U>>(f: F) -> impl SliceIo {
+    FnWrapper {
+        f,
+        phantom: PhantomData,
+    }
+}
+
 struct FnWrapper<T: Pod, U: Pod, F: Fn(&[T]) -> Vec<U>> {
     f: F,
     phantom: PhantomData<(T, U)>,
 }
 
-impl<T: Pod, U: Pod, F: Fn(&[T]) -> Vec<U>> SliceIoHandler for FnWrapper<T, U, F> {
+impl<T: Pod, U: Pod, F: Fn(&[T]) -> Vec<U>> SliceIo for FnWrapper<T, U, F> {
     type FromGuest = T;
     type ToGuest = U;
 
-    fn handle_io(&self, from_guest: &[Self::FromGuest]) -> Vec<Self::ToGuest> {
+    fn handle_io(&self, _syscall: &str, from_guest: &[Self::FromGuest]) -> Vec<Self::ToGuest> {
         (self.f)(from_guest)
     }
 }
+
+/// Posix-style IO
+pub(crate) struct PosixIo<'a> {
+    read_fds: RefCell<BTreeMap<u32, Rc<RefCell<dyn BufRead + 'a>>>>,
+    write_fds: RefCell<BTreeMap<u32, Rc<RefCell<dyn Write + 'a>>>>,
+}
+
+impl<'a> PosixIo<'a> {
+    pub fn new() -> Self {
+        Self {
+            read_fds: RefCell::new(BTreeMap::new()),
+            write_fds: RefCell::new(BTreeMap::new()),
+        }
+    }
+
+    pub fn with_read_fd(self, fd: u32, reader: Rc<RefCell<impl BufRead + 'a>>) -> Self {
+        self.read_fds.borrow_mut().insert(fd, reader);
+        self
+    }
```

Comment:
I remember you mentioned needing a `RefCell` in the platform sync, but I don't see why it's needed here at least, this is an owned reference. I see in the `Syscall::syscall` method it needs some mutability. Could that be solved by using a `&mut self` receiver. I'm guessing I'm missing something. Since this exhibits observable mutability, I agree that it would be nice to make that clear in the type system, but that `RefCell` is a decent backup option.